hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/ServerCall.java

```diff
@@ -209,8 +213,24 @@ public RequestHeader getHeader() {
   }
 
   @Override
-  public RPCProtos.ConnectionHeader getConnectionHeader() {
-    return this.connection.connectionHeader;
+  public Map<String, byte[]> getConnectionAttributes() {
+    return this.connection.connectionAttributes;
+  }
+
+  @Override
+  public Map<String, byte[]> getRequestAttributes() {
+    if (this.requestAttributes == null) {
+      if (header.getAttributeList().isEmpty()) {
+        this.requestAttributes = Collections.emptyMap();
+      } else {
+        this.requestAttributes = Maps.newHashMapWithExpectedSize(header.getAttributeList().size());
```

Comment:
I don't think we need to synchronize this method given the chances of threading issues (low) and the cost of the computation (low), but we should consider thread safety here nonetheless. 

At the very least, we should make `this.requestAttributes` volatile. Then here we should construct a local map and add the values onto it, then only do `this.requestAttributes = localAttributes` at the end. This way if there did happen to be 2 callers, one would not get an incomplete view of the request attributes.